spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -45,36 +45,61 @@
  */
 public class PubSubTemplate implements PubSubOperations, InitializingBean {
 
+	private static final String DEFAULT_SOURCE_NAME = "spring";
+
 	private final String projectId;
 
 	private final GoogleCredentials credentials;
 
 	private ConcurrentHashMap<String, Publisher> publishers = new ConcurrentHashMap<>();
 
+	/**
+	 * Aimed at reusing the same executor across every {@link Publisher}, so the number of allocated
+	 * threads doesn't blow out of proportion.
+	 */
 	private ExecutorProvider executorProvider;
 
 	private MessageConverter messageConverter = new SimpleMessageConverter();
 
+	/**
+	 * The Spring component a Google Cloud Pub/Sub call is originated from.
+	 *
+	 * <p>Spring Integration callers should set this to "spring-integration", Spring Cloud to
+	 * "spring-cloud", etc.
+	 */
+	private String sourceName = DEFAULT_SOURCE_NAME;
+
+	/**
+	 * The version of the Spring component a Google Cloud Pub/Sub call is originated from.
+	 *
+	 * <p>By default, the version is obtained from this package's manifest file, which is mirrored
+	 * from this package's pom.xml file. Ideally, this would be the version of the library invoking
+	 * this call (e.g., Spring Integration, Spring Cloud, etc.).
+	 */
+	private String sourceVersion;
+
 	private int concurrentProducers = 1;
 
 	public PubSubTemplate(GoogleCredentials credentials, String projectId) {
 		this.projectId = projectId;
 		this.credentials = credentials;
+		this.sourceVersion = this.getClass().getPackage().getImplementationVersion();
 		this.executorProvider = InstantiatingExecutorProvider.newBuilder()
 				.setExecutorThreadCount(this.concurrentProducers).build();
 	}
 
 	@Override
 	public ListenableFuture<String> send(final String topic, Message message) {
-
-		Publisher publisher = this.publishers.computeIfAbsent(topic, s -> {
+		// Reuse existing publisher if one already exists.
+		Publisher publisher = this.publishers.computeIfAbsent(topic, key -> {
 			try {
-				return Publisher.defaultBuilder(TopicName.create(this.projectId, topic))
+				return Publisher.defaultBuilder(TopicName.create(this.projectId, key))
 						.setExecutorProvider(this.executorProvider)
```

Comment:
@artembilan Correct me if I'm not understanding correctly, but the whole concern of creating a `Publisher` should be delegated to a Publisher factory, which in turn would depend on a `ChannelProvider`, and calling the `setClientLibHeader` would be the responsibility of whatever is creating the `ChannelProvider`. Basically, the whole creation of the `Publisher` should be done through dependency injection.